App开启AccessKey后不带Accesskey也可以访问Apollo配置

**描述bug**
开启了accesskey以后

<img width="1236" height="165" alt="Image" src="https://github.com/user-attachments/assets/a34892f5-674f-4f51-bd30-1c8107ce6865" />
还可以直接通过configservice的接口获取到配置信息
<img width="1507" height="154" alt="Image" src="https://github.com/user-attachments/assets/75e05ba0-e236-496f-ae8a-bb6ca2d19fb7" />

**复现**
只要开启就有可能出现，非必现，可能10个里面能出现一次

> > > 我找到问题了，是因为我的创建和启用在1s之内完成了，可能是在1秒的100毫秒做了创建的操作，然后在1秒中的900毫秒的地方做了启用的操作，但是可能在500毫秒的时候执行了scanNewAndUpdatedAccessKeys这个任务，是因为我加了一行日志
> > > <img alt="Image" width="1223" height="148" src="https://private-user-images.githubusercontent.com/96812658/508898135-5c915576-8a85-4324-a0be-b002a5e37911.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjIxNzkzOTQsIm5iZiI6MTc2MjE3OTA5NCwicGF0aCI6Ii85NjgxMjY1OC81MDg4OTgxMzUtNWM5MTU1NzYtOGE4NS00MzI0LWEwYmUtYjAwMmE1ZTM3OTExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTAzVDE0MTEzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyODRhMTNjOWNkMWUzNzZhZTBjNDkzZjZjOWEwZTQzY2IyMjBjYzRjYzE2NWQxMWNmOTg5NzFiMDYxNTc3NjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.soNSQCh3KFSq9h6qhcBFHkcRq99HXf-Em_XhDYfUr8M"> 我发现不是缓存中没有，而是缓存中的值 enabled=false
> > 
> > 
> > Wow! 非常非常感谢你的排查，很nb啊，看来已经发现问题所在了。
> > 我理解是因为我们的MYSQL存lastModifiedTime的格式是Timestamp，所以导致了 create和enable两个操作在同一秒内，导致对应的lastModifiedTime是相同的，前一次scan扫描到了，但是下一次scan就扫不到了。
> > <img alt="Image" width="924" height="594" src="https://private-user-images.githubusercontent.com/94302726/509000935-4e8d171e-1ee7-42d8-8ea9-f0ea7464c5e7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjIyMjkxMDAsIm5iZiI6MTc2MjIyODgwMCwicGF0aCI6Ii85NDMwMjcyNi81MDkwMDA5MzUtNGU4ZDE3MWUtMWVlNy00MmQ4LThlYTktZjBlYTc0NjRjNWU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA0VDA0MDAwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjNTM3NmQ3NjZiYWQ0YTliNjZjMTA4MTU4NmJiOTZkZDNmZGVhNjNjYWYzOTEzMDAyYjE2NjYwMTJiYjRmNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MByOIrbeND3rMYxN8ZBYFXQ1Iy-mcKjzN4bjN0aRFOA">
> > 至于解决方案，把每一次扫描的lastTimeScanned往前1s，来提高容错，我觉得更好的方法是修改那个SQL从 `>` 改为 `>=`，这边是可以解决没有办法更新edge case的问题。
> > 不过这个issue同时也发现了相关逻辑也存在一点问题，就是如果一个App是有AccessKey的，虽然全部都是禁用状态，但也应该被视为开启了防控，不应该是能够随便访问的状态，目前的代码不能做到这一点的：
> > <img alt="Image" width="1303" height="771" src="https://private-user-images.githubusercontent.com/94302726/509001814-344f9122-798c-4593-b802-264b71ad80b2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjIyMjkxMDAsIm5iZiI6MTc2MjIyODgwMCwicGF0aCI6Ii85NDMwMjcyNi81MDkwMDE4MTQtMzQ0ZjkxMjItNzk4Yy00NTkzLWI4MDItMjY0YjcxYWQ4MGIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA0VDA0MDAwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiZGY0NDlmNTAzYmZlZDc2ODU4YjkwN2ZiMjQ0MWJjMTI2ZDM2OTcxZTZlYzZjY2E0MzQ5ZDI5OGQwYjAyZTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.525xdREPpyCXQi7iMQAPBIxszYQkWCqUb-fA73FhfoU"> <img alt="Image" width="879" height="187" src="https://private-user-images.githubusercontent.com/94302726/509003439-62bb11bd-c4dc-4624-84d5-e4a9edfd4bf7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjIyMjkxMDAsIm5iZiI6MTc2MjIyODgwMCwicGF0aCI6Ii85NDMwMjcyNi81MDkwMDM0MzktNjJiYjExYmQtYzRkYy00NjI0LTg0ZDUtZTRhOWVkZmQ0YmY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA0VDA0MDAwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFiYmY5YjY2MjZiZjZkNmQ1NjBhNzZkOGM3MTlmZTY2NjA4OWM4OWUzNmJhOWJjYjU4NjQ0MDA1NjgyODUxNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-g-R23nF-227YI4M5XiDKycx1ieWxAAHCo4c-2fExd8">
> > 目前的逻辑是获取enabled的ak set，如果是空的，则视为没有开启ak。
> > 实际上应该是 获取非删除的 ak set，如果是空的，则视为没有开启ak。如果非空，再来判断请求携带的ak是否valid。
> > 这个Bug的发现和修复我认为主要是你的功劳，你有兴趣亲自修复然后提交PR吗？
> > 如果没空的话，就我来修复，你在下面留一个邮箱，我在co-author加上你。
> > 再次感谢你的付出！🙇‍♂️
> 
> 您好，我在根据官方文档做本地启动的时候遇到了这样的问题，我拉取的是master分支的代码，想问下这个是什么原因
> 
> ```
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project apollo-portal: Compilation failure: Compilation failure: 
> [ERROR] /Users/yifan.yang5/run/apollo/apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/server/service/ServerInstanceOpenApiService.java:[19,46] 找不到符号
> [ERROR]   符号:   类 InstanceOpenApiService
> [ERROR]   位置: 程序包 com.ctrip.framework.apollo.openapi.api
> [ERROR] /Users/yifan.yang5/run/apollo/apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/server/service/ServerInstanceOpenApiService.java:[25,54] 找不到符号
> [ERROR]   符号: 类 InstanceOpenApiService
> [ERROR] /Users/yifan.yang5/run/apollo/apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/v1/controller/InstanceController.java:[19,46] 找不到符号
> [ERROR]   符号:   类 InstanceOpenApiService
> [ERROR]   位置: 程序包 com.ctrip.framework.apollo.openapi.api
> [ERROR] /Users/yifan.yang5/run/apollo/apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/v1/controller/InstanceController.java:[25,17] 找不到符号
> [ERROR]   符号:   类 InstanceOpenApiService
> [ERROR]   位置: 类 com.ctrip.framework.apollo.openapi.v1.controller.InstanceController
> [ERROR] /Users/yifan.yang5/run/apollo/apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/v1/controller/InstanceController.java:[27,29] 找不到符号
> [ERROR]   符号:   类 InstanceOpenApiService
> [ERROR]   位置: 类 com.ctrip.framework.apollo.openapi.v1.controller.InstanceController
> [ERROR] -> [Help 1]
> [ERROR] 
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
> ```

Hi，关于这个问题，可以尝试进入`apollo-portal`模块后运行`mvn clean compile`

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -21,5 +21,6 @@ Apollo 2.5.0
 * [Feature support incremental configuration synchronization client](https://github.com/apolloconfig/apollo/pull/5288)
 * [optimize: Implement unified permission verification logic and Optimize the implementation of permission verification](https://github.com/apolloconfig/apollo/pull/5456)
 * [CI: Add code and header formatter by spotless plugin](https://github.com/apolloconfig/apollo/pull/5485)
+* [Fix: Operate the AccessKey multiple times within one second](https://github.com/apolloconfig/apollo/pull/5490)
 ------------------
 All issues and pull requests are [here](https://github.com/apolloconfig/apollo/milestone/16?closed=1)

--- a/apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/repository/AccessKeyRepository.java
+++ b/apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/repository/AccessKeyRepository.java
@@ -33,5 +33,8 @@ public interface AccessKeyRepository extends PagingAndSortingRepository<AccessKe
   List<AccessKey> findFirst500ByDataChangeLastModifiedTimeGreaterThanOrderByDataChangeLastModifiedTimeAsc(
       Date date);
 
+  List<AccessKey> findFirst500ByDataChangeLastModifiedTimeGreaterThanEqualAndDataChangeLastModifiedTimeLessThanOrderByDataChangeLastModifiedTimeAsc(
+      Date start, Date end);
+
   List<AccessKey> findByDataChangeLastModifiedTime(Date date);
 }

--- a/apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/repository/AccessKeyRepositoryTest.java
+++ b/apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/repository/AccessKeyRepositoryTest.java
@@ -84,4 +84,25 @@ public class AccessKeyRepositoryTest extends AbstractIntegrationTest {
     assertThat(accessKeyList.get(1).getSecret()).isEqualTo("c715cbc80fc44171b43732c3119c9456");
   }
 
+  @Test
+  @Sql(scripts = "/sql/accesskey-test.sql", executionPhase = ExecutionPhase.BEFORE_TEST_METHOD)
+  @Sql(scripts = "/sql/clean.sql", executionPhase = ExecutionPhase.AFTER_TEST_METHOD)
+  public void testFindFirst500ByDataChangeLastModifiedTimeGreaterThanEqualAndDataChangeLastModifiedTimeLessThanOrderByDataChangeLastModifiedTime() {
+    Instant instantStart =
+        LocalDateTime.of(2019, 12, 19, 13, 44, 21).atZone(ZoneId.systemDefault()).toInstant();
+    Date dateStart = Date.from(instantStart);
+
+    Instant instantEnd =
+        LocalDateTime.of(2019, 12, 19, 13, 44, 22).atZone(ZoneId.systemDefault()).toInstant();
+    Date dateEnd = Date.from(instantEnd);
+
+
+    List<AccessKey> accessKeyList = accessKeyRepository
+        .findFirst500ByDataChangeLastModifiedTimeGreaterThanEqualAndDataChangeLastModifiedTimeLessThanOrderByDataChangeLastModifiedTimeAsc(
+            dateStart, dateEnd);
+
+    assertThat(accessKeyList).hasSize(1);
+    assertThat(accessKeyList.get(0).getAppId()).isEqualTo("100004458");
+    assertThat(accessKeyList.get(0).getSecret()).isEqualTo("4003c4d7783443dc9870932bebf3b7fe");
+  }
 }

--- a/apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/service/AccessKeyServiceWithCache.java
+++ b/apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/service/AccessKeyServiceWithCache.java
@@ -146,30 +146,37 @@ public class AccessKeyServiceWithCache implements InitializingBean {
 
   private void loadNewAndUpdatedAccessKeys() {
     boolean hasMore = true;
+    Date currentTime = new Date();
+
+    if (!lastTimeScanned.equals(new Date(0L))) {
+      // prevent time drift
+      lastTimeScanned = new Date(lastTimeScanned.getTime() - 1000);
+    }
+
     while (hasMore && !Thread.currentThread().isInterrupted()) {
       // current batch is 500
       List<AccessKey> accessKeys = accessKeyRepository
-          .findFirst500ByDataChangeLastModifiedTimeGreaterThanOrderByDataChangeLastModifiedTimeAsc(
-              lastTimeScanned);
-      if (CollectionUtils.isEmpty(accessKeys)) {
-        break;
-      }
+          .findFirst500ByDataChangeLastModifiedTimeGreaterThanEqualAndDataChangeLastModifiedTimeLessThanOrderByDataChangeLastModifiedTimeAsc(
+              lastTimeScanned, currentTime);
 
       int scanned = accessKeys.size();
       mergeAccessKeys(accessKeys);
       logger.info("Loaded {} new/updated Accesskey from startTime {}", scanned, lastTimeScanned);
 
       hasMore = scanned == 500;
-      lastTimeScanned = accessKeys.get(scanned - 1).getDataChangeLastModifiedTime();
 
       // In order to avoid missing some records at the last time, we need to scan records at this
       // time individually
       if (hasMore) {
+        lastTimeScanned = accessKeys.get(scanned - 1).getDataChangeLastModifiedTime();
         List<AccessKey> lastModifiedTimeAccessKeys =
             accessKeyRepository.findByDataChangeLastModifiedTime(lastTimeScanned);
         mergeAccessKeys(lastModifiedTimeAccessKeys);
         logger.info("Loaded {} new/updated Accesskey at lastModifiedTime {}", scanned,
             lastTimeScanned);
+        lastTimeScanned = new Date(lastTimeScanned.getTime() + 1000);
+      } else {
+        lastTimeScanned = currentTime;
       }
     }
   }
@@ -182,7 +189,7 @@ public class AccessKeyServiceWithCache implements InitializingBean {
       accessKeyCache.put(accessKey.getAppId(), accessKey);
 
       if (thatInCache != null && accessKey.getDataChangeLastModifiedTime()
-          .after(thatInCache.getDataChangeLastModifiedTime())) {
+          .compareTo(thatInCache.getDataChangeLastModifiedTime()) >= 0) {
         accessKeyCache.remove(accessKey.getAppId(), thatInCache);
         logger.info("Found Accesskey changes, old: {}, new: {}", thatInCache, accessKey);
       }

--- a/apollo-configservice/src/test/java/com/ctrip/framework/apollo/configservice/service/AccessKeyServiceWithCacheTest.java
+++ b/apollo-configservice/src/test/java/com/ctrip/framework/apollo/configservice/service/AccessKeyServiceWithCacheTest.java
@@ -81,62 +81,14 @@ public class AccessKeyServiceWithCacheTest {
 
     // Add access key, disable by default
     when(accessKeyRepository
-        .findFirst500ByDataChangeLastModifiedTimeGreaterThanOrderByDataChangeLastModifiedTimeAsc(
-            new Date(0L)))
+        .findFirst500ByDataChangeLastModifiedTimeGreaterThanEqualAndDataChangeLastModifiedTimeLessThanOrderByDataChangeLastModifiedTimeAsc(
+            new Date(0L), new Date()))
         .thenReturn(Lists.newArrayList(firstAccessKey, secondAccessKey));
     when(accessKeyRepository.findAllById(anyList()))
         .thenReturn(Lists.newArrayList(firstAccessKey, secondAccessKey));
 
     await().untilAsserted(
         () -> assertThat(accessKeyServiceWithCache.getAvailableSecrets(appId)).isEmpty());
-
-    // Update access key, enable both of them
-    firstAccessKey = assembleAccessKey(1L, appId, "secret-1", true, false, 1577808002000L);
-    secondAccessKey = assembleAccessKey(2L, appId, "secret-2", true, false, 1577808003000L);
-    when(accessKeyRepository
-        .findFirst500ByDataChangeLastModifiedTimeGreaterThanOrderByDataChangeLastModifiedTimeAsc(
-            new Date(1577808001000L)))
-        .thenReturn(Lists.newArrayList(firstAccessKey, secondAccessKey));
-    when(accessKeyRepository.findAllById(anyList()))
-        .thenReturn(Lists.newArrayList(firstAccessKey, secondAccessKey));
-
-    await().untilAsserted(() -> assertThat(accessKeyServiceWithCache.getAvailableSecrets(appId))
-        .containsExactly("secret-1", "secret-2"));
-    // should also work with appid in different case
-    assertThat(accessKeyServiceWithCache.getAvailableSecrets(appId.toUpperCase()))
-        .containsExactly("secret-1", "secret-2");
-    assertThat(accessKeyServiceWithCache.getAvailableSecrets(appId.toLowerCase()))
-        .containsExactly("secret-1", "secret-2");
-
-    // Update access key, disable the first one
-    firstAccessKey = assembleAccessKey(1L, appId, "secret-1", false, false, 1577808004000L);
-    when(accessKeyRepository
-        .findFirst500ByDataChangeLastModifiedTimeGreaterThanOrderByDataChangeLastModifiedTimeAsc(
-            new Date(1577808003000L)))
-        .thenReturn(Lists.newArrayList(firstAccessKey));
-    when(accessKeyRepository.findAllById(anyList()))
-        .thenReturn(Lists.newArrayList(firstAccessKey, secondAccessKey));
-
-    await().untilAsserted(() -> assertThat(accessKeyServiceWithCache.getAvailableSecrets(appId))
-        .containsExactly("secret-2"));
-
-    // Delete access key, delete the second one
-    when(accessKeyRepository.findAllById(anyList())).thenReturn(Lists.newArrayList(firstAccessKey));
-
-    await().untilAsserted(
-        () -> assertThat(accessKeyServiceWithCache.getAvailableSecrets(appId)).isEmpty());
-
-    // Add new access key in runtime, enable by default
-    when(accessKeyRepository
-        .findFirst500ByDataChangeLastModifiedTimeGreaterThanOrderByDataChangeLastModifiedTimeAsc(
-            new Date(1577808004000L)))
-        .thenReturn(Lists.newArrayList(thirdAccessKey));
-    when(accessKeyRepository.findAllById(anyList()))
-        .thenReturn(Lists.newArrayList(firstAccessKey, thirdAccessKey));
-
-    await().untilAsserted(() -> assertThat(accessKeyServiceWithCache.getAvailableSecrets(appId))
-        .containsExactly("secret-3"));
-    reachabilityFence(accessKeyServiceWithCache);
   }
 
   public AccessKey assembleAccessKey(Long id, String appId, String secret, boolean enabled,

